src/librustc_parse/parser/pat.rs

```diff
@@ -670,59 +608,66 @@ impl<'a> Parser<'a> {
+    /// `$begin $form` has already been parsed.
+    fn parse_pat_range_begin_with(
+        &mut self,
+        begin: P<Expr>,
+        re: Spanned<RangeEnd>,
+    ) -> PResult<'a, PatKind> {
+        let end = if self.is_pat_range_end_start(0) {
+            // Parsing e.g. `X..=Y`.
+            Some(self.parse_pat_range_end()?)
+        } else {
+            // Parsing e.g. `X..`.
+            self.sess.gated_spans.gate(sym::half_open_range_patterns, begin.span.to(re.span));
+            if let RangeEnd::Included(_) = re.node {
+                // FIXME(Centril): Consider semantic errors instead in `ast_validation`.
+                // Possibly also do this for `X..=` in *expression* contexts.
+                self.error_inclusive_range_with_no_end(re.span);
+            }
+            None
+        };
+        Ok(PatKind::Range(Some(begin), end, re))
+    }
 
-        self.diagnostic()
-            .struct_span_err(range_span, &format!("`{}X` range patterns are not supported", form))
-            .span_suggestion(
-                range_span,
-                "try using the minimum value for the type",
-                format!("MIN{}{}", form, pprust::expr_to_string(&end)),
-                Applicability::HasPlaceholders,
-            )
+    pub(super) fn error_inclusive_range_with_no_end(&self, span: Span) {
+        use rustc_error_codes::E0586;
+        struct_span_err!(self.sess.span_diagnostic, span, E0586, "inclusive range with no end")
+            .help("inclusive ranges must be bounded at the end (`..=b` or `a..=b`)")
             .emit();
-
-        Ok(PatKind::Range(begin, end, respan(lo, re)))
     }
 
-    /// Parse the end of a `X..Y`, `X..=Y`, or `X...Y` range pattern  or recover
-    /// if that end is missing treating it as `X..`, `X..=`, or `X...` respectively.
-    fn parse_pat_range_end_opt(&mut self, begin: &Expr, form: &str) -> PResult<'a, P<Expr>> {
-        if self.is_pat_range_end_start() {
-            // Parsing e.g. `X..=Y`.
-            self.parse_pat_range_end()
-        } else {
-            // Parsing e.g. `X..`.
-            let range_span = begin.span.to(self.prev_span);
-
-            self.diagnostic()
-                .struct_span_err(
-                    range_span,
-                    &format!("`X{}` range patterns are not supported", form),
-                )
-                .span_suggestion(
-                    range_span,
-                    "try using the maximum value for the type",
-                    format!("{}{}MAX", pprust::expr_to_string(&begin), form),
-                    Applicability::HasPlaceholders,
-                )
-                .emit();
+    /// Parse a range-to pattern, e.g. `..X` and `..=X` where `X` remains to be parsed.
+    fn parse_pat_range_to(&mut self, re: Spanned<RangeEnd>) -> PResult<'a, PatKind> {
+        let end = self.parse_pat_range_end()?;
+        self.sess.gated_spans.gate(sym::half_open_range_patterns, re.span.to(self.prev_span));
```

Comment:
@cramertj I can fairly easily do:
```suggestion
        self.sess.gated_spans.gate(sym::half_open_range_patterns, re.span.to(self.prev_span));
        if re.node == RangeEnd::Included(RangeSyntax::DotDotDot) {
            re.node = RangeEnd::Included(RangeSyntax::DotDotEq);
            self.struct_span_err(re.span, "range-to patterns with `...` are not allowed")
                 .span_suggestion_hidden(
                     re.span,
                     "use `..=` instead",
                     "..=".to_string(),
                     Applicability::MachineApplicable,
                 )
                 .emit();
        }
```

and this will specifically disallow `...X` (but with recovery and everything...) and keep our options open.
The cost is of course a slightly less consistent grammar, but it's a fairly minor hit.

Reply:
(Another alternative would be to do this check in AST validation or lowering to keep the grammar more consistent.)

Reply:
@cramertj Ping on this :)

Reply:
Filed follow-up in https://github.com/rust-lang/rust/pull/68120.